include/swift/Driver/Compilation.h

```diff
@@ -90,6 +90,11 @@ class Compilation {
   /// The translated input arg list.
   std::unique_ptr<llvm::opt::DerivedArgList> TranslatedArgs;
 
+  /// The ToolChain, which the Compilation takes ownership of upon
+  /// construction, and which it may use to build subsequent batch
+  /// jobs itself.
+  std::unique_ptr<const ToolChain> TheToolChain;
```

Comment:
This seems wrong to me. One Driver should (in theory) be able to perform multiple Compilations with the same ToolChain. I think that's more important than having Compilation stand on its own without a Driver.

(Also, I don't think `ToolChain` makes a const/non-const distinction to begin with, so it would be better to omit `const` here.)

Reply:
@jrose-apple Do you mean, by this, that I should copy the toolchain into the Compilation, by value? Or make it a shared pointer?

To be clear about the motive here: the issue isn't that I want there to be any particular numerical relationship between the number of ToolChains and Compilations in a given run; the issue is just that the ToolChain doesn't currently escape Driver::buildCompilation -- Compilation outlives ToolChain altogether -- and I need to be able to recover it (or "some" ToolChain) in order to re-enter Toolchain::constructInvocation and its callees, when building up a BatchJob, from within Compilation::performJobsImpl's callees.

I suppose I could also just re-manufacture a ToolChain from Driver anytime I need one, it .. looks like all the pieces are lying around for that to work. Would that be better?

Reply:
I assumed the Driver owned both the ToolChain and the Compilation (well, by virtue of owning the stack), and therefore making ToolChain a reference here would be fine. A shared pointer would also be fine with me.

The original model for Compilation vs. Driver was that you could build a Compilation and execute it later, not that anyone would do that. (The thing you might do is build a Compilation, ask questions about it, and throw it away.) That's a little muddied now that the Compilation isn't "fully built" ahead of time.

None of this *really* matters. It's just object graph mental models.

Reply:
(P.S. You can't copy the ToolChain by value because it's a virtual class hierarchy.)

Reply:
Ok, so you're suggesting I move the lifetime of Toolchain up the stack, to encompass the lifetime of Compilation, and then have Compilation hold a reference? That's fine by me (and I'll proceed that way until I hear otherwise here, trying not to waste your time too much!)